nautobot/utilities/filters.py

```diff
@@ -286,6 +286,141 @@ def filter(self, qs, value):
         return qs
 
 
+class MappedPredicatesFilterMixin:
+    """
+    A filter mixin to provide the ability to specify fields and lookup expressions to use for
+    filtering.
+
+    A mapping of filter predicates (field_name: lookup_expr) must be provided to the filter when
+    declared on a filterset. This mapping is used to construct a `Q` query to filter based on the
+    provided predicates.
+
+    Example:
+
+        q = SearchFilter(
+            filter_predicates={
+                "comments": "icontains",
+                "name": "icontains",
+                "id": "iexact",
+            },
+        )
+
+    Optionally you may also provide a callable to use as a preprocessor for the filter predicate by
+    providing the value as a nested dict with "lookup_expr" and "preprocessor" keys. For example:
+
+        q = SearchFilter(
+            filter_predicates={
+                "asn": {
+                    "lookup_expr": "exact",
+                    "preprocessor": int,
+            },
+        )
+
+    This tells the filter to try to cast `asn` to an `int`. If it fails, this predicate will be
+    skipped.
+    """
+
+    # Optional label for the form element generated for this filter
+    label = None
+
+    def __init__(self, filter_predicates=None, label=None, *args, **kwargs):
+        if not isinstance(filter_predicates, dict):
+            raise TypeError("filter_predicates must be a dict")
+
+        # Format: {field_name: lookup_expr, ...}
+        self.filter_predicates = filter_predicates
+
+        # Try to use the label from the class if it is defined.
+        if label is None:
+            kwargs.setdefault("label", self.label)
+
+        super().__init__(*args, **kwargs)
+
+        # Generate the query with a sentinel value to validate it and surface parse errors.
+        self.generate_query(self.filter_predicates, value="")
+
+    def generate_query(self, filter_predicates, value):
+        """
+        Given a mapping of `filter_predicates` and a `value`, return a `Q` object for 2-tuple of
+        predicate=value.
+        """
+        query = models.Q()
+        for field_name, lookup_info in filter_predicates.items():
+            if isinstance(lookup_info, str):
+                lookup_expr = lookup_info
+                preprocessor = str.strip
+            elif isinstance(lookup_info, dict):
+                lookup_expr = lookup_info["lookup_expr"]
+                preprocessor = lookup_info.get("preprocessor")
```

Comment:
I had it that way before and changed it to just let `preprocessor` be `None` and fail. In this current form, if one uses the dictionary form for specifying the lookup_expr but not also the preprocessor, then it should immediately fail. Thus the check for `if not callable(preprocessor)`, it would disallow this to be optional in the event that you've used this form:

```{"foo": {"lookup_expr": "exact"}```

This is the same as `{"foo": "exact"}` but defaults to using `str.strip` as the preprocessor. I am of the mind that in the expanded case, NOT specifying `preprocessor` is an error (because it will default to `None` which is not callable and fill raise a `TypeError`. So it's either use the default `{field_name: lookup_expr}` form and let it fall back to `str.strip) OR use `field_name: `{"lookup_expr": lookup_expr, "preprocessor": my_callabale}`  but not both.

Reply:
Fine by me.

Reply:
This has been implemented as of the latest pushes.